File: src/hw1/task3_4/Card.java
package hw1.task3_4;

public class Card implements Comparable<Card> {
    private Rank rank;
    private Suit suit;

    public Card(Rank rank, Suit suit) {
        this.rank = rank;
        this.suit = suit;
    }

    public Rank getRank() {
        return this.rank;
    }

    public Suit getSuit() {
        return this.suit;
    }

    @Override
    public String toString() {
        return rank + " " + suit;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Card)) {
            return false;
        }
        Card other = (Card) obj;
        return rank == other.getRank() && suit.equals(other.getSuit());
    }

    @Override
    public int hashCode() {
        return rank.getRankValue() * 3 + suit.toString().length() * 5;
    }

    public boolean isCardOfStandardDeck() {
        return rank.getRankValue() >= 2 && rank.getRankValue() <= 14 && suit != null;
    }

    public boolean isStronger(Card card) {
        return suit.equals(card.getSuit()) && rank.getRankValue() > card.getRank().getRankValue();
    }

    public static int compare(Card card1, Card card2) {
        return card1.compareTo(card2);
    }

    @Override
    public int compareTo(Card card) {
        int suitWeight = suit.getSuitValue();
        int otherSuitWeight = card.getSuit().getSuitValue();
        if (equals(card)){
            return 0;
        } else if ((suitWeight == otherSuitWeight && !isStronger(card)) || (suitWeight < otherSuitWeight)){
            return -1;
        } else {
            return 1;
        }
    }
}

File: src/hw2/task3/Card.kt
package hw2.task3

class Card(private val rank: Rank, private val suit: Suit) : Comparable<Card> {

    override fun toString(): String {
        return "$rank $suit"
    }

    override fun equals(other: Any?): Boolean {
        if (other ==null) return false
        if (other !is Card) {
            return false
        }
        if (this === other) return true
        return rank == other.rank && suit == other.suit
    }

    override fun hashCode(): Int = rank.rankValue * 3 + suit.toString().length * 5

    fun isCardOfStandardDeck(): Boolean {
        return rank.rankValue in 2..14
    }

    private fun isStronger(card : Card): Boolean {
        return suit == card.suit && rank.rankValue > card.rank.rankValue
    }

    fun compare(card1: Card, card2: Card) : Int = card1.compareTo(card2)

    override fun compareTo(other: Card): Int {
        val suitWeight = suit.suitValue
        val otherSuitWeight = other.suit.suitValue
        return if (equals(other)) 0
        else if (suitWeight == otherSuitWeight && !isStronger(other) || (suitWeight < otherSuitWeight)) -1
        else 1
    }
}